### PR TITLE
Feature/named credentials

### PR DESCRIPTION
**Pull request by dfruddffdc**
Migrating Named Credential support from internal repo

### CODE REVIEW

File: src/classes/ffhttp_AbstractFormDataSerializable.cls-meta.xml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<ApexClass xmlns="http://soap.sforce.com/2006/04/metadata">
+    <apiVersion>38.0</apiVersion>
```
Comment (mwoodcertinia) [inline src/classes/ffhttp_AbstractFormDataSerializable.cls-meta.xml:3]: Think we should look at updating the apiVersion for all meta data files

File: src/classes/ffhttp_AbstractFormDataSerializable.cls [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+/**
+ * ffhttp_AbstractFormDataSerializable 
+ * 
+ * Abstract class to allow the custom serialisation of form encoded objects.
+ * This class extends IHttpContent which allows these objects to be sent in HTTP requests.
+ */
+public abstract class ffhttp_AbstractFormDataSerializable implements ffhttp_IHttpContent
+{
+	private List < String > m_parameters = new List < String > ();
```
Comment (mwoodcertinia) [inline src/classes/ffhttp_AbstractFormDataSerializable.cls]: Beautify has caused issues with lists.
Can we remove the extra spaces.

File: src/classes/ffhttp_Client.cls [hunk 3]
```diff
@@ -29,86 +29,310 @@
...
+	private static final String CURL_HEADER = ' -H ';
+	private static final String CURL_BODY = ' -d ';
+	private static final String CURL_SPACE_SEPARATOR = ' ';
+	private static final String CURL_QUOTE_SEPARATOR = '"';
+	private static final String CURL_COLON_SEPARATOR = ':';
+
 	/**
 	 * Credentials Interface
 	 */
 	public interface ICredentials
```
Comment (mwoodcertinia) [inline src/classes/ffhttp_Client.cls:186]: Add something to the readme as this is a breaking change.
I also think we should create a change log.

File: src/classes/ffhttp_AbstractFormDataSerializableTest.cls [hunk 4]
```diff
@@ -0,0 +1,71 @@
+/**
```
Comment (mwoodcertinia) [inline src/classes/ffhttp_AbstractFormDataSerializableTest.cls:1]: Can we do a test deploy and make sure that all these tests pass.

Comment (dfruddffdc) [reply src/classes/ffhttp_AbstractFormDataSerializableTest.cls:1]: Yes, I did this already, tests passed and coverage was good.

File: src/classes/ffhttp_AbstractFormDataSerializable.cls-meta.xml (hunk 1 above)
Comment (dfruddffdc) [reply src/classes/ffhttp_AbstractFormDataSerializable.cls-meta.xml:3]: Is that in the scope of this change?

File: src/classes/ffhttp_AbstractFormDataSerializable.cls (hunk 2 above)
Comment (dfruddffdc) [reply src/classes/ffhttp_AbstractFormDataSerializable.cls]: That would make me happy, I don't use it for Apex anyway :)

File: src/classes/ffhttp_AbstractFormDataSerializable.cls-meta.xml (hunk 1 above)
Comment (dfruddffdc) [reply src/classes/ffhttp_AbstractFormDataSerializable.cls-meta.xml:3]: Leaving for now, as discussed.

File: src/classes/ffhttp_AbstractFormDataSerializable.cls (hunk 2 above)
Comment (dfruddffdc) [reply src/classes/ffhttp_AbstractFormDataSerializable.cls]: Done

File: src/classes/ffhttp_Client.cls (hunk 3 above)
Comment (dfruddffdc) [reply src/classes/ffhttp_Client.cls:186]: https://github.com/financialforcedev/ffhttp-core/blob/feature/named-credentials/CHANGELOG.md

**Review by mwoodcertinia — APPROVED** [review]
Happy for this to be merged.

**Pull request merged**